base/array.jl

```diff
@@ -207,9 +207,15 @@ fill(v, dims::Integer...) = fill!(Array{typeof(v)}(dims...), v)
 
 for (fname, felt) in ((:zeros,:zero), (:ones,:one))
     @eval begin
-        ($fname)(T::Type, dims...)       = fill!(Array{T}(dims...), ($felt)(T))
-        ($fname)(dims...)                = fill!(Array{Float64}(dims...), ($felt)(Float64))
-        ($fname){T}(A::AbstractArray{T}) = fill!(similar(A), ($felt)(T))
+        function ($fname)(a::AbstractArray, T::Type=eltype(a), dims::Tuple=size(a))
+            fill!(similar(a,T,dims), $felt(T))
+        end
+        ($fname)(T::Type, dims::Tuple) = ($fname)(Array{T}(dims...), T, dims)
```

Comment:
This isn't right: it allocates an `Array{T}(dims...)` and passes it to `$fname(...)`, but then that function calls `similar`, which allocates an *second* array.   So you end up allocating one more array than you need.

I think you just want
```jl
($fname)(T::Type, dims::Tuple) = fill!(Array{T}(dims...), $felt(T))
```